mitmproxy/protocol/socks_client.py

```diff
@@ -0,0 +1,82 @@
+import socket
+
+from netlib import socks
+from netlib.exceptions import TcpException
+from netlib import tcp
+
+from .base import Layer
+from ..exceptions import ProtocolException
+
+class SocksClientLayer(Layer):
+
+    def __init__(self, ctx, server_tls):
+        Layer.__init__(self, ctx)
+        self.server_tls = server_tls
+        self.socks_addr = tcp.Address(("127.0.0.1", 1080))
+        self.socks_username = "test"
+        self.socks_password = "test"
+
+    def __call__(self):
+        layer = self.ctx.next_layer(self)
+        layer()
+
+    def _socks_username_password_auth(self, rfile, wfile):
+        if self.socks_username is None or self.socks_password is None:
```

Comment:
This check is not duplicate.When the username and password for MITMProxy not set and also the SOCKS server tell us we can only accept METHOD.USERNAME_PASSWORD,  it can throw an exception. :wink: